File: backend/app/services/inspection_service.py
"""Service layer for inspection operations."""

from datetime import datetime, timezone
from typing import List, Optional
from uuid import UUID

from sqlalchemy.exc import SQLAlchemyError
from sqlalchemy.orm import Session

from ..database import SessionLocal
from ..models.inspection import Inspection, InspectionStatus


class InspectionService:
    """Service class for managing inspection operations."""

    def create_inspection(
        self,
        building_name: str,
        building_address: str,
        operator_name: str,
        total_area_sqm: float,
        inspection_date: Optional[datetime] = None
    ) -> Inspection:
        """
        Create a new inspection in the database.

        Args:
            building_name: Name of the building being inspected
            building_address: Address of the building
            operator_name: Name of the operator performing the inspection
            total_area_sqm: Total area of the building in square meters
            inspection_date: Date of the inspection (defaults to current time if not provided)

        Returns:
            The created Inspection object

        Raises:
            ValueError: If validation fails (e.g., negative area)
            SQLAlchemyError: If database operation fails
        """
        if total_area_sqm <= 0:
            raise ValueError("total_area_sqm must be positive")

        if inspection_date is None:
            inspection_date = datetime.now(timezone.utc)

        session: Session = SessionLocal()
        try:
            inspection = Inspection(
                building_name=building_name,
                building_address=building_address,
                operator_name=operator_name,
                total_area_sqm=total_area_sqm,
                inspection_date=inspection_date,
                status=InspectionStatus.PENDING
            )
            session.add(inspection)
            session.commit()
            session.refresh(inspection)
            return inspection
        except SQLAlchemyError as e:
            session.rollback()
            raise SQLAlchemyError(f"Failed to create inspection: {str(e)}") from e
        finally:
            session.close()

    def get_inspection(self, inspection_id: UUID) -> Optional[Inspection]:
        """
        Retrieve an inspection by its ID.

        Args:
            inspection_id: UUID of the inspection to retrieve

        Returns:
            The Inspection object if found, None otherwise

        Raises:
            SQLAlchemyError: If database operation fails
        """
        session: Session = SessionLocal()
        try:
            inspection = session.query(Inspection).filter(
                Inspection.id == inspection_id
            ).first()
            return inspection
        except SQLAlchemyError as e:
            raise SQLAlchemyError(f"Failed to retrieve inspection: {str(e)}") from e
        finally:
            session.close()

    def list_inspections(
        self,
        skip: int = 0,
        limit: int = 100
    ) -> List[Inspection]:
        """
        Retrieve a paginated list of inspections.

        Args:
            skip: Number of records to skip (for pagination)
            limit: Maximum number of records to return

        Returns:
            List of Inspection objects ordered by created_at descending

        Raises:
            SQLAlchemyError: If database operation fails
        """
        session: Session = SessionLocal()
        try:
            inspections = session.query(Inspection).order_by(
                Inspection.created_at.desc()
            ).offset(skip).limit(limit).all()
            return inspections
        except SQLAlchemyError as e:
            raise SQLAlchemyError(f"Failed to list inspections: {str(e)}") from e
        finally:
            session.close()

    def update_inspection_status(
        self,
        inspection_id: UUID,
        status: str
    ) -> Optional[Inspection]:
        """
        Update the status of an inspection.

        Args:
            inspection_id: UUID of the inspection to update
            status: New status value (must be valid InspectionStatus enum value)

        Returns:
            The updated Inspection object if found, None if inspection not found

        Raises:
            ValueError: If status is not a valid enum value
            SQLAlchemyError: If database operation fails
        """
        # Validate status is a valid enum value
        try:
            status_enum = InspectionStatus(status)
        except ValueError:
            valid_values = [s.value for s in InspectionStatus]
            raise ValueError(
                f"Invalid status '{status}'. Must be one of: {valid_values}"
            )

        session: Session = SessionLocal()
        try:
            inspection = session.query(Inspection).filter(
                Inspection.id == inspection_id
            ).first()

            if inspection is None:
                return None

            inspection.status = status_enum
            session.commit()
            session.refresh(inspection)
            return inspection
        except SQLAlchemyError as e:
            session.rollback()
            raise SQLAlchemyError(f"Failed to update inspection status: {str(e)}") from e
        finally:
            session.close()

    def update_quality_score(
        self,
        inspection_id: UUID,
        score: float
    ) -> Optional[Inspection]:
        """
        Update the overall quality score of an inspection.

        Args:
            inspection_id: UUID of the inspection to update
            score: Quality score (must be between 0 and 100)

        Returns:
            The updated Inspection object if found, None if inspection not found

        Raises:
            ValueError: If score is not between 0 and 100
            SQLAlchemyError: If database operation fails
        """
        # Validate score range
        if score < 0 or score > 100:
            raise ValueError(
                f"Quality score must be between 0 and 100, got {score}"
            )

        session: Session = SessionLocal()
        try:
            inspection = session.query(Inspection).filter(
                Inspection.id == inspection_id
            ).first()

            if inspection is None:
                return None

            inspection.overall_quality_score = score
            session.commit()
            session.refresh(inspection)
            return inspection
        except SQLAlchemyError as e:
            session.rollback()
            raise SQLAlchemyError(f"Failed to update quality score: {str(e)}") from e
        finally:
            session.close()

File: frontend/src/services/inspectionService.js
/**
 * Inspection Service - API functions for managing building inspections
 *
 * This module provides functions to interact with the inspections API endpoints.
 * All functions use async/await and proper error handling.
 */

import { apiClient } from '../config/api.js';

/**
 * Create a new inspection
 *
 * @param {Object} data - Inspection data
 * @param {string} data.building_name - Name of the building (required)
 * @param {string} data.building_address - Address of the building (required)
 * @param {string} data.operator_name - Name of the operator (required)
 * @param {number} data.total_area_sqm - Total area in square meters (required, must be positive)
 * @returns {Promise<Object>} Created inspection object
 * @throws {Error} If validation fails or request fails
 *
 * @example
 * const inspection = await createInspection({
 *   building_name: "Test Tower",
 *   building_address: "Dubai Marina",
 *   operator_name: "Ashish",
 *   total_area_sqm: 1000.0
 * });
 */
export async function createInspection(data) {
  try {
    const response = await apiClient.post('/inspections', data);
    return response.data;
  } catch (error) {
    // Extract meaningful error message
    const errorMessage = error.response?.data?.detail
      || error.response?.data?.message
      || error.message
      || 'Failed to create inspection';

    throw new Error(errorMessage);
  }
}

/**
 * Get inspection by ID
 *
 * @param {string} id - UUID of the inspection
 * @returns {Promise<Object>} Inspection object
 * @throws {Error} If inspection not found or request fails
 *
 * @example
 * const inspection = await getInspection('4950262f-f3bb-41ad-873d-add5b10534c4');
 */
export async function getInspection(id) {
  try {
    const response = await apiClient.get(`/inspections/${id}`);
    return response.data;
  } catch (error) {
    // Handle 404 specifically
    if (error.response?.status === 404) {
      throw new Error(`Inspection with id ${id} not found`);
    }

    // Extract meaningful error message
    const errorMessage = error.response?.data?.detail
      || error.response?.data?.message
      || error.message
      || 'Failed to get inspection';

    throw new Error(errorMessage);
  }
}

/**
 * List inspections with pagination
 *
 * @param {number} skip - Number of records to skip (default: 0)
 * @param {number} limit - Maximum number of records to return (default: 100)
 * @returns {Promise<Array>} Array of inspection objects (empty array if none found)
 * @throws {Error} If request fails
 *
 * @example
 * const inspections = await listInspections(0, 10);
 */
export async function listInspections(skip = 0, limit = 100) {
  try {
    const response = await apiClient.get('/inspections', {
      params: { skip, limit }
    });
    return response.data;
  } catch (error) {
    // On error, return empty array instead of mock data
    console.error('Failed to list inspections:', error.message);
    return [];
  }
}

/**
 * Update inspection status
 *
 * @param {string} id - UUID of the inspection
 * @param {string} status - New status value (pending, in_progress, completed, failed)
 * @returns {Promise<Object>} Updated inspection object
 * @throws {Error} If inspection not found, invalid status, or request fails
 *
 * @example
 * const updated = await updateInspectionStatus('4950262f-f3bb-41ad-873d-add5b10534c4', 'completed');
 */
export async function updateInspectionStatus(id, status) {
  try {
    const response = await apiClient.patch(`/inspections/${id}/status`, {
      status
    });
    return response.data;
  } catch (error) {
    // Handle 404 specifically
    if (error.response?.status === 404) {
      throw new Error(`Inspection with id ${id} not found`);
    }

    // Handle validation errors
    if (error.response?.status === 422 || error.response?.status === 400) {
      const detail = error.response?.data?.detail;
      if (Array.isArray(detail)) {
        // Pydantic validation error
        throw new Error(detail[0]?.msg || 'Invalid status value');
      } else if (typeof detail === 'string') {
        throw new Error(detail);
      }
    }

    // Extract meaningful error message
    const errorMessage = error.response?.data?.detail
      || error.response?.data?.message
      || error.message
      || 'Failed to update inspection status';

    throw new Error(errorMessage);
  }
}

/**
 * Update inspection quality score
 *
 * @param {string} id - UUID of the inspection
 * @param {number} score - Quality score between 0 and 100
 * @returns {Promise<Object>} Updated inspection object
 * @throws {Error} If inspection not found, invalid score, or request fails
 *
 * @example
 * const updated = await updateInspectionScore('4950262f-f3bb-41ad-873d-add5b10534c4', 85.5);
 */
export async function updateInspectionScore(id, score) {
  try {
    const response = await apiClient.patch(`/inspections/${id}/score`, {
      score
    });
    return response.data;
  } catch (error) {
    // Handle 404 specifically
    if (error.response?.status === 404) {
      throw new Error(`Inspection with id ${id} not found`);
    }

    // Handle validation errors
    if (error.response?.status === 422 || error.response?.status === 400) {
      const detail = error.response?.data?.detail;
      if (Array.isArray(detail)) {
        // Pydantic validation error
        throw new Error(detail[0]?.msg || 'Invalid score value');
      } else if (typeof detail === 'string') {
        throw new Error(detail);
      }
    }

    // Extract meaningful error message
    const errorMessage = error.response?.data?.detail
      || error.response?.data?.message
      || error.message
      || 'Failed to update inspection score';

    throw new Error(errorMessage);
  }
}
